pkg/cachemanager/cachemanager_test/cachemanager_integration_test.go

```diff
@@ -0,0 +1,333 @@
+)
+
+var cfg *rest.Config
+
+func TestMain(m *testing.M) {
+	testutils.StartControlPlane(m, &cfg, 3)
+}
+
+// TestCacheManager_replay_retries tests that we can retry GVKs that error out in the reply goroutine.
+func TestCacheManager_replay_retries(t *testing.T) {
+	mgr, wm := testutils.SetupManager(t, cfg)
+	c := testclient.NewRetryClient(mgr.GetClient())
+
+	failPlease := make(chan string, 2)
+	reader := fakes.HookReader{
+		Reader: c,
+		ListFunc: func(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+			// Return an error the first go-around.
+			var failKind string
+			select {
+			case failKind = <-failPlease:
+			default:
+			}
+			if failKind != "" && list.GetObjectKind().GroupVersionKind().Kind == failKind {
+				return fmt.Errorf("synthetic failure")
+			}
+			return c.List(ctx, list, opts...)
+		},
+	}
+
+	cacheManager, dataStore, _, ctx := cacheManagerForTest(t, mgr, wm, reader)
+
+	cfClient, ok := dataStore.(*cachemanager.FakeCfClient)
+	require.True(t, ok)
+
+	// seed one gvk
+	configMapGVK := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "ConfigMap"}
+	cm := unstructuredFor(configMapGVK, "config-test-1")
+	require.NoError(t, c.Create(ctx, cm), "creating ConfigMap config-test-1")
+
+	podGVK := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Pod"}
+	pod := unstructuredFor(podGVK, "pod-1")
+	require.NoError(t, c.Create(ctx, pod), "creating Pod pod-1")
+
+	syncSourceOne := aggregator.Key{Source: "source_a", ID: "ID_a"}
+	require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{configMapGVK, podGVK}))
+
+	expected := map[cachemanager.CfDataKey]interface{}{
+		{Gvk: configMapGVK, Key: "default/config-test-1"}: nil,
+		{Gvk: podGVK, Key: "default/pod-1"}:               nil,
+	}
+
+	require.Eventually(t, expectedCheck(cfClient, expected), eventuallyTimeout, eventuallyTicker)
+
+	// set up a scenario where the list from replay will fail a few times
+	failPlease <- "ConfigMapList"
+	failPlease <- "ConfigMapList"
+
+	// this call should schedule a cache wipe and a replay for the configMapGVK
+	require.NoError(t, cacheManager.UpsertSource(ctx, syncSourceOne, []schema.GroupVersionKind{configMapGVK}))
+
+	expected2 := map[cachemanager.CfDataKey]interface{}{
+		{Gvk: configMapGVK, Key: "default/config-test-1"}: nil,
+	}
+	require.Eventually(t, expectedCheck(cfClient, expected2), eventuallyTimeout, eventuallyTicker)
+
+	// cleanup
+	require.NoError(t, c.Delete(ctx, cm), "creating ConfigMap config-test-1")
+	require.NoError(t, c.Delete(ctx, pod), "creating ConfigMap pod-1")
+}
+
+// TestCacheManager_concurrent makes sure that we can add and remove multiple sources
+// from separate go routines and changes to the underlying cache are reflected.
+func TestCacheManager_concurrent(t *testing.T) {
+	mgr, wm := testutils.SetupManager(t, cfg)
+	c := testclient.NewRetryClient(mgr.GetClient())
+	cacheManager, dataStore, agg, ctx := cacheManagerForTest(t, mgr, wm, c)
```

Comment:
Might be nice to have this return a struct, since it has so many values.  That also saves you from having to do things like

```
var1, _, var2, _ := ...
```

To get the code to compile.  

Something like `testResources` that's package private.

Reply:
OTOH, then every future access to a variable requires something like `testResources.var1` to be accessed (or a separate variable assignment).

Biggest benefit of using a struct is not needing to remember the semantic significance of index Y in a return tuple, and the ability to add new return values in a backwards-compatible manner.

Not intending to veto this comment, but figured I'd add some nuance.

Reply:
I opted w going w the `testResources`